packages/edit-widgets/src/store/batch-processing/reducer.js

```diff
@@ -0,0 +1,221 @@
+/**
+ * External dependencies
+ */
+import { omit } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import {
+	BATCH_MAX_SIZE,
+	STATE_NEW,
+	STATE_IN_PROGRESS,
+	STATE_SUCCESS,
+	STATE_ERROR,
+} from './constants';
+
+const defaultState = {
+	lastBatchId: 0,
+	enqueuedItems: {},
+	batches: {},
+	processors: {},
+	promises: {},
+};
+
+export default function reducer( state = defaultState, action ) {
+	switch ( action.type ) {
+		case 'ENQUEUE_ITEM': {
+			const { queue, context, item, itemId } = action;
+
+			const stateQueue = state.enqueuedItems[ queue ] || {};
+			const stateItems = stateQueue[ context ] || [];
+
+			return {
+				...state,
+				enqueuedItems: {
+					...state.enqueuedItems,
+					[ queue ]: {
+						...stateQueue,
+						[ context ]: [ ...stateItems, { id: itemId, item } ],
+					},
+				},
+			};
+		}
+
+		case 'PREPARE_BATCH_FOR_PROCESSING': {
+			const { queue, context, batchId, meta } = action;
+
+			if ( batchId in state.batches ) {
+				throw new Error( `Batch ${ batchId } already exists` );
+			}
+
+			const stateQueue = state.enqueuedItems[ queue ] || {};
+			const enqueuedItems = [ ...( stateQueue[ context ] || [] ) ];
+			const transactions = {};
+			let transactionNb = 0;
+			while ( enqueuedItems.length ) {
+				const transactionId = `${ batchId }-${ transactionNb }`;
+				transactions[ transactionId ] = {
+					number: transactionNb,
+					id: transactionId,
+					items: enqueuedItems.splice( 0, BATCH_MAX_SIZE ),
+				};
+				++transactionNb;
+			}
+
+			const batch = {
+				id: batchId,
+				state: STATE_NEW,
+				queue,
+				context,
+				transactions,
+				results: {},
+				meta,
+			};
+
+			return {
+				...state,
+				enqueuedItems: {
+					...state.enqueuedItems,
+					[ queue ]: {
+						...stateQueue,
+						[ context ]: [],
+					},
+				},
+				batches: {
+					...state.batches,
+					[ batchId ]: batch,
+				},
+			};
+		}
+
+		case 'SETUP_PROMISE': {
+			return {
+				...state,
+				promises: {
+					...state.promises,
+					[ action.queue ]: {
+						...( state.promises[ action.queue ] || {} ),
+						[ action.context ]: {
+							promise: action.promise,
+							resolve: action.resolve,
+							reject: action.reject,
+						},
+					},
+				},
+			};
+		}
```

Comment:
Promises are mutable but redux state should be immutable by definition - let's remove this in a follow up PR.

Reply:
So putting this in just a global somewhere?

Reply:
I was thinking more about not storing this anywhere, only returning it to the caller. It's not a very well-thought proposal, I understand you added that for a reason. Let's figure it out after 9.2 RC1 - this works and is fine for now.